### DISCUSSION TITLE
How to check that branch is already merged

### DISCUSSION THREAD

**Discussion by kazhuravlev** [go-git/go-git#951] [Q&A] [answered]
I would like to get list of branches that already merged like `git branch --merged` or not merged yet `git branch --no-merged`. 

I tried to workaround with `object.Tree.Diff`, but it always have a diff (that reasonable). Probably I should read all target (ex. master) branch commits and try to find a merge commits?

**Comment by kazhuravlev** [comment]
The main problem is that I don't fully understand the internals of git. I think I can find **some** way, but I want to understand the **true** way.

**Reply by pjbgf** [reply, marked as answer]
For `--merged`, I believe you would get the tip commit of your branches, then iterate through the commits of your current branch to confirm whether they can be found there. 

For getting tip commits from the branches you could use something like this:

```go
    branches, err := r.Branches()
    ...

    err = branches.ForEach(func(b *plumbing.Reference) error {
        // Skip the current branch
        if b.Name() == headRef.Name() {
            return nil
        }
        ...
        commit, err := r.CommitObject(b.Hash())
```

For iterating through the commits from the current branch, you could use `r.Log`:

```go
    headRef, err := r.Head() // Tip of the current branch
    ...
    cIter, err := r.Log(&git.LogOptions{From: headRef.Hash})
```

Or navigating the parents of `r.Head()`.

**Reply by kazhuravlev** [reply]
Yes, it works! Thank you for fast response)
